Framework/DataHandling/src/LoadCSNSNexus.cpp

```diff
@@ -0,0 +1,568 @@
+ */
+void LoadCSNSNexus::init() {
+  declareProperty(make_unique<PropertyWithValue<string>>("Instrument", "GPPD",
+                                                         Direction::Input),
+                  "Different instrument with different detector combinations");
+
+  const vector<string> exts{".h5", ".nxs"};
+  declareProperty(
+      make_unique<FileProperty>("Filename", "", FileProperty::Load, exts),
+      "The name of the Nexus file to load");
+
+  declareProperty(std::make_unique<PropertyWithValue<string>>(
+                      "NXentryName", "csns", Direction::Input),
+                  "Optional: Name of entry (default csns)");
+
+  declareProperty(
+      make_unique<PropertyWithValue<bool>>("LoadBank", true, Direction::Input),
+      "Default true: load bank data, false: load monitor data.");
+
+  declareProperty(
+      std::make_unique<ArrayProperty<string>>("Bankname", Direction::Input),
+      "Optional: A comma-separated list of bank/monitor to read");
+
+  declareProperty(
+      std::make_unique<ArrayProperty<uint32_t>>("StartT0", Direction::Input),
+      "Optional: A comma-separated list of StartNo of T0 to read.");
+  declareProperty(
+      std::make_unique<ArrayProperty<uint32_t>>("EndT0", Direction::Input),
+      "Optional: A comma-separated list of endNo of T0 to read.");
+
+  declareProperty(make_unique<PropertyWithValue<bool>>("LoadEvent", false,
+                                                       Direction::Input),
+                  "Default false: load event data, true: load histogram data.");
+
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "An output orkspace.");
+
+  std::string grp1 = "Bank/Monitor";
+  setPropertyGroup("Bankname", grp1);
+  setPropertyGroup("loadBank", grp1);
+
+  std::string grp2 = "FilterByPulse";
+  setPropertyGroup("LoadEvent", grp2);
+  setPropertyGroup("StartT0", grp2);
+  setPropertyGroup("EndT0", grp2);
+}
+
+/**
+ * Return the start or end time of experiment
+ * @param[in] typeName :: start_time_utc or end_time_utc
+ * @return expTime
+ */
+Types::Core::DateAndTime LoadCSNSNexus::getExperimentTime(string typeName) {
+  Types::Core::DateAndTime data;
+  std::string expTime;
+  m_file->openGroup(m_entry, "NXentry");
+  std::map<std::string, std::string> entries = m_file->getEntries();
+  if (entries.find(typeName) != entries.end())
+
+    m_file->readData(typeName, expTime);
+  else
+    throw std::runtime_error("No experiment time in the NXS file!");
+  m_file->closeGroup();
+  try {
+    data = Mantid::Types::Core::DateAndTime(expTime);
+  } catch (...) {
+    data = Mantid::Types::Core::DateAndTime("2100-12-31T00:00:00+08:00");
+  }
+  return data;
+}
+
+/**
+ * Return modules for special bankName at GPPD
+ * @param[in] bankName :: bank1,bank2,bank3
+ * @return modules
+ */
+vector<string> LoadCSNSNexus::getGPPDModules(string bankName) {
+  vector<string> data;
```

Comment:
In C++11 `auto` is useful for writing things like this instead of creating the variable then assigning to it:
```
auto entries = m_file->getEntries();
```
There are also range-based and structured bindings in C++17 that make iteration over a `map` much more readable.

There is a fair bit of code duplication here. Could we write this as:

```c++
int firstModuleSuffix(-1), secondModuleSuffix(-1);
if(bankName == "bank3") {
  firstModuleId = 1;
  secondModuleId = 2;
} else if(bankName == "bank2") {
  firstModuleId = 3;
  secondModuleId = 4;
} else if(bankName == "bank1") {
  firstModuleId = 5;
  secondModuleId = 6;
} else {
  return;
}

vector<string> data;
m_file->openGroup(m_entry, "NXentry");
m_file->openGroup("instrument", "NXinstrument");
for (const auto & [name,_] : entries) {
  if (name.compare(0, 7, "module" + std::to_string(firstModuleId)) == 0 ||
      name.compare(0, 7, "module" + std::to_string(secondModuleId)) == 0) {
    data.emplace_back(name);
  }
} 
m_file->closeGroup();
m_file->closeGroup();
return data;
```

Reply:
`for (const auto & [name,_] : entries) {`
It has a warning :unused variable ‘_’ [-Wunused-variable] for (const auto &[name, _] : entries) { 
And it failed in 'Ubuntu + Doc Tests'.
Can I change this line to 
`for (auto it: entries) {
          std::string name = it.first;`

Reply:
Thanks for pointing this out. This is an issue with structured bindings that will be addressed in future. Yes your suggestion is fine but we should remember to use references to avoid copying:
```c++
for (const auto & it: entries) { const auto & name = it.first;
```